File: ibkr_show_backend/app/schemas/positions.py
from pydantic import BaseModel

from app.schemas.common import PaginationInfo


class PositionItem(BaseModel):
    account_id: str
    report_date: str
    symbol: str | None = None
    description: str | None = None
    asset_class: str | None = None
    quantity: float | None = None
    mark_price: float | None = None
    position_value: float | None = None
    percent_of_nav: float | None = None
    average_cost_price: float | None = None
    cost_basis_money: float | None = None
    total_realized_pnl: float | None = None
    realized_pnl_percent: float | None = None
    total_unrealized_pnl: float | None = None
    unrealized_pnl_percent: float | None = None
    total_fifo_pnl: float | None = None
    previous_day_change_percent: float | None = None


class PositionListResponse(BaseModel):
    items: list[PositionItem]
    pagination: PaginationInfo


class PositionDetailBar(BaseModel):
    report_date: str
    open_price: float | None = None
    high_price: float | None = None
    low_price: float | None = None
    close_price: float | None = None
    quantity: float | None = None


class PositionDetailTradeMarker(BaseModel):
    trade_date: str | None = None
    date_time: str | None = None
    buy_sell: str | None = None
    quantity: float | None = None
    trade_price: float | None = None
    fifo_pnl_realized: float | None = None


class PositionDetailResponse(BaseModel):
    symbol: str | None = None
    description: str | None = None
    asset_class: str | None = None
    bars: list[PositionDetailBar]
    trades: list[PositionDetailTradeMarker]


class PositionConcentrationItem(BaseModel):
    symbol: str | None = None
    description: str | None = None
    asset_class: str | None = None
    position_value: float = 0.0
    percent_of_nav: float | None = None


class PositionAssetDistributionItem(BaseModel):
    asset_class: str | None = None
    position_value: float = 0.0
    positions_count: int = 0


class PositionSummaryResponse(BaseModel):
    report_date: str | None = None
    total_positions: int = 0
    total_position_value: float = 0.0
    total_cost_basis_money: float = 0.0
    total_realized_pnl: float = 0.0
    total_unrealized_pnl: float = 0.0
    total_fifo_pnl: float = 0.0
    top_positions: list[PositionConcentrationItem]
    asset_distribution: list[PositionAssetDistributionItem]

File: ibkr_show_frontend/src/types/positions.ts
import type { PaginationInfo } from './common'

export interface PositionItem {
  account_id: string
  report_date: string
  symbol: string | null
  description: string | null
  asset_class: string | null
  quantity: number | null
  mark_price: number | null
  position_value: number | null
  percent_of_nav: number | null
  average_cost_price: number | null
  cost_basis_money: number | null
  total_realized_pnl: number | null
  realized_pnl_percent: number | null
  total_unrealized_pnl: number | null
  unrealized_pnl_percent: number | null
  total_fifo_pnl: number | null
  previous_day_change_percent: number | null
}

export interface PositionListResponse {
  items: PositionItem[]
  pagination: PaginationInfo
}

export interface PositionConcentrationItem {
  symbol: string | null
  description: string | null
  asset_class: string | null
  position_value: number
  percent_of_nav: number | null
}

export interface PositionAssetDistributionItem {
  asset_class: string | null
  position_value: number
  positions_count: number
}

export interface PositionSummaryResponse {
  report_date: string | null
  total_positions: number
  total_position_value: number
  total_cost_basis_money: number
  total_realized_pnl: number
  total_unrealized_pnl: number
  total_fifo_pnl: number
  top_positions: PositionConcentrationItem[]
  asset_distribution: PositionAssetDistributionItem[]
}

export interface PositionDetailBar {
  report_date: string
  open_price: number | null
  high_price: number | null
  low_price: number | null
  close_price: number | null
  quantity: number | null
}

export interface PositionDetailTradeMarker {
  trade_date: string | null
  date_time: string | null
  buy_sell: string | null
  quantity: number | null
  trade_price: number | null
  fifo_pnl_realized: number | null
}

export interface PositionDetailResponse {
  symbol: string | null
  description: string | null
  asset_class: string | null
  bars: PositionDetailBar[]
  trades: PositionDetailTradeMarker[]
}
